### ISSUE TITLE
some theme items do not change colors when requested

### ISSUE THREAD

**Issue by fdncred** [nushell/nushell#9264]
### Describe the bug

I'm trying to write a script that will:
1. load a theme
2. display the theme colors
3. automatically take a screenshot and save it
4. clear the screen
5. rinse and repeat

What I've noticed is that colors for headers (and maybe other things) do not change colors when requested via a theme change.

### preview-theme.nu script
```bash
def preview_theme [] {
    let ansi_names = (ansi --list | get name)
    let all_ansi_names = $ansi_names ++ (ansi --list | get 'short name' | range 133..388)
    let color_config = ($env.config.color_config | transpose key value)
    let color_table = ($color_config | each {|row|
      if ($row.value | describe | str contains 'closure') {
        # get the closure as a string
        let source_code = (view source ($env.config.color_config | get $row.key) | str replace -a "'" '')
        # replace named colors with ansi codes
        let source_code_replaced = ($all_ansi_names | reduce -f $source_code {|it, acc| $acc | str replace $"\\b($it)\\b" $"(ansi $it)($it)(ansi reset)"})
        # todo, how do we color the closure hex colors?
        [[key value]; [$row.key $source_code_replaced]]
      } else if ($row.value | describe | str contains 'record') {
        [[key value]; [$row.key $"(ansi ($row.value))($row.value)(ansi reset)"]]
      } else if ($row.value | str starts-with '#') {
        [[key value]; [$row.key $"(ansi ($row.value))($row.value)(ansi reset)"]]
      } else {
        [[key value]; [$row.key $"(ansi ($row.value))($row.value)(ansi reset)"]]
      }
    } | flatten)

    let row_count = ($color_table | length)
    let row_count_half = (($color_table | length) / 2 | math floor)
    let table1 = ($color_table | range 0..$row_count_half | rename datatypes dtvals)
    let table2 = ($color_table | range $row_count_half..$row_count | rename shapes shpvals)

    echo $table1 | merge $table2
}
```

### preview-generate-script.nu (a script to generate a script)
```bash
let themes = (ls *.nu | first 10)
$themes | each {|theme|
  let name = $theme.name
  let basename = ($theme.name | path parse | get stem)
  echo "clear\n" | save -f -a script.nu
  echo $"use ($name)\n" | save -f -a script.nu
  echo $"$env.config.color_config = (char lp)($basename)(char rp)\n" | save -f -a script.nu
  # echo "source preview-current.nu\n" | save -f -a script.nu
  echo "source preview-theme.nu\n" | save -f -a script.nu
  echo "sleep 100ms\n" | save -f -a script.nu
  echo $"print (char sq)($basename)(char sq)\n" | save -f -a script.nu
  # echo "print (main)\n" | save -f -a script.nu
  echo "print (preview_theme)\n" | save -f -a script.nu
  echo "sleep 250ms\n" | save -f -a script.nu
  echo $'c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_($name | path parse | get stem)_$date$.png -closeapp -exit(char nl)' | save -f -a script.nu
}
```

### preview-screenshot-script.nu
```bash
clear
use 3024-day.nu
$env.config.color_config = (3024-day)
source preview-theme.nu
sleep 100ms
print '3024-day'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_3024-day_$date$.png -closeapp -exit
clear
use 3024-night.nu
$env.config.color_config = (3024-night)
source preview-theme.nu
sleep 100ms
print '3024-night'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_3024-night_$date$.png -closeapp -exit
clear
use 3024.nu
$env.config.color_config = (3024)
source preview-theme.nu
sleep 100ms
print '3024'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_3024_$date$.png -closeapp -exit
clear
use abyss.nu
$env.config.color_config = (abyss)
source preview-theme.nu
sleep 100ms
print 'abyss'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_abyss_$date$.png -closeapp -exit
clear
use aci.nu
$env.config.color_config = (aci)
source preview-theme.nu
sleep 100ms
print 'aci'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_aci_$date$.png -closeapp -exit
clear
use aco.nu
$env.config.color_config = (aco)
source preview-theme.nu
sleep 100ms
print 'aco'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_aco_$date$.png -closeapp -exit
clear
use adventuretime.nu
$env.config.color_config = (adventuretime)
source preview-theme.nu
sleep 100ms
print 'adventuretime'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_adventuretime_$date$.png -closeapp -exit
clear
use afterglow.nu
$env.config.color_config = (afterglow)
source preview-theme.nu
sleep 100ms
print 'afterglow'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_afterglow_$date$.png -closeapp -exit
clear
use alien-blood.nu
$env.config.color_config = (alien-blood)
source preview-theme.nu
sleep 100ms
print 'alien-blood'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_alien-blood_$date$.png -closeapp -exit
clear
use alucard.nu
$env.config.color_config = (alucard)
source preview-theme.nu
sleep 100ms
print 'alucard'
print (preview_theme)
sleep 250ms
c:\apps\MiniCap\MiniCap.exe -captureactivewin -bordershadow -save c:\temp\theme_alucard_$date$.png -closeapp -exit
```

### How to reproduce

The screenshot script only works on windows right now.

1. clone nu_scripts repo
2. cd to nu_scripts/themes/themes
3. copy the scripts above into this directory

To preview your current theme:
1. source preview-theme.nu
2. preview_theme

To generate the screenshots from the current directory (this only does the first 10)
1. preview-screenshot-script.nu

Examples of what it's currently generating (yes, it's not working correctly)
### 3024
![theme_3024_5_22_2023](https://github.com/nushell/nushell/assets/343840/5598af28-bde3-4224-8982-c119346c6b8f)
### 3024-day
![theme_3024_5_22_2023](https://github.com/nushell/nushell/assets/343840/2ae73d79-a8ac-49bd-9b17-e0fffb1d4068)
### 3024-night
![theme_3024-night_5_22_2023](https://github.com/nushell/nushell/assets/343840/40df3557-7b6a-49c5-9bc0-ca649cb6ce26)
### abyss
![theme_abyss_5_22_2023](https://github.com/nushell/nushell/assets/343840/a5c776e6-1a62-46e0-a5b6-6eae03c8a9b4)
### aci
![theme_aci_5_22_2023](https://github.com/nushell/nushell/assets/343840/7cd76f1a-0589-475e-9d3d-0629b7af21d8)
### aco
![theme_aco_5_22_2023](https://github.com/nushell/nushell/assets/343840/70c0cb94-079a-472a-a9c3-d37501a669ed)
### adventuretime
![theme_adventuretime_5_22_2023](https://github.com/nushell/nushell/assets/343840/fbc39454-7fdd-4820-955c-2e736421207e)
### afterglow
![theme_afterglow_5_22_2023](https://github.com/nushell/nushell/assets/343840/d5675cc7-a2ff-419b-b2a9-d48b0326cf22)
### alien-blood
![theme_alien-blood_5_22_2023](https://github.com/nushell/nushell/assets/343840/fb37c45b-4ac3-49b6-80f8-3c58e6fc4f1c)
### alucard
![theme_alucard_5_22_2023](https://github.com/nushell/nushell/assets/343840/c70218ce-bfd6-4c33-8b9f-427e8102cd53)

Notice the column headers and row headers do not change colors. I'm not sure why. I'm guessing a bug in our code somewhere.

### Expected behavior

All theme colors to be shown correctly.

### Configuration

|key|value|
|-|-|
|version|0.80.1|
|branch|main|
|commit_hash|057de0661315e0b4e6f38aafca13179a4d9e8369|
|build_os|windows-x86_64|
|build_target|x86_64-pc-windows-msvc|
|rust_version|rustc 1.67.1 (d5a82bbd2 2023-02-07)|
|rust_channel|1.67.1-x86_64-pc-windows-msvc|
|cargo_version|cargo 1.67.1 (8ecd4f20a 2023-01-10)|
|build_time|2023-05-17 11:19:17 -05:00|
|build_rust_channel|release|
|features|default, zip|
|installed_plugins|from eml, from ics, from ini, from parquet, from vcf, gstat, hist, inc, json path, plot, query, query json, query web, query xml, regex, xyplot|

**Comment by zhiburt** [comment]
Hi @fdncred 

You've put quite an effort into  the scripts;

Seems to work for me.

I don't know maybe something in your scripts?
Could you use this instead and verify? (you must enter it twice to see effect)

```nu
use alucard.nu ; $env.config.color_config = (alucard); preview_theme
```

![image](https://github.com/nushell/nushell/assets/20165848/98ad3aae-f36d-4ddb-b9fc-75d78f75cedd)

![image](https://github.com/nushell/nushell/assets/20165848/00cdbb0c-1a3e-4c0b-9a4f-2bb0ddd10f1c)

![image](https://github.com/nushell/nushell/assets/20165848/d5f7eadf-95cc-4b71-888a-8cdf55efcc78)

![image](https://github.com/nushell/nushell/assets/20165848/5311da04-b0c3-441f-a2bb-a0a71b47409b)

_______________________________________

PS: I did not figured out how to make it a function, seems like `use` requires a constant?

![image](https://github.com/nushell/nushell/assets/20165848/b30e3711-ac0d-4518-93a9-90fc29090e4f)

**Comment by fdncred** [comment]
@zhiburt I've got updates to the script but it doesn't affect this problem. You're right, that if you run them one at a time in the repl, it seems to work, although sometimes you have to run it twice.

However, if you run the script above `preview-screenshot-script.nu` (just comment out the c:\apps\MiniCap\MiniCap.exe part), which switches themes in a script, you should see the problem.

**Comment by zhiburt** [comment]
So I guess I see what you mean.

I think it's a shorter example. (though it has a different cause I guess).

```nu
def ctxIssue [] { $env.config.color_config.header = blue; ls } 
ctxIssue
```

Seems like `merge_delta`/`merge_env` must be done for blocks, calls? (`nu_engine/src/eval.rs`).

I've approached it a number of times but yet not succeed.
As it's a bit tricky cause being called recursively.

I might be wrong though.

**Comment by zhiburt** [comment]
Yes seems like I am correct;

At least what worked for me;

For functions you need to remove a callee stack (or merge it later).
And then update the engine before print is called (not it updates it only after).

https://github.com/nushell/nushell/blob/74724dee809c039440ce299db65ee5ad5aa321ea/crates/nu-cli/src/util.rs#L241-L245

https://github.com/nushell/nushell/blob/74724dee809c039440ce299db65ee5ad5aa321ea/crates/nu-engine/src/eval.rs#L59-L59

I do not provide a PR as I am not sure about the changes.
I very likely it can be done differently.

**Comment by fdncred** [comment]
@zhiburt This patch doesn't seem to fix it when running in a script for me. The header, row_index, and separators are all colored whatever they were with the last theme. Since my default theme has cyan headers, yellow row_index, and gold separators, so do all themes ran from the script. although you can look at the tables and see what they're supposed to be.
![theme_aco_5_24_2023](https://github.com/nushell/nushell/assets/343840/5c424883-89f3-4781-8245-33a2616f3ec1)
![theme_alien-blood_5_24_2023](https://github.com/nushell/nushell/assets/343840/0de814d7-fdc9-451e-9678-38c3bde7e855)
![theme_alucard_5_24_2023](https://github.com/nushell/nushell/assets/343840/1d84a544-543a-497f-8a19-5a21d5068257)
![theme_adventuretime_5_24_2023](https://github.com/nushell/nushell/assets/343840/c5216570-7435-4b44-9631-36fe5d707f25)

If you comment out the C:\apps\minicap\minicap.exe line in the preview-screenshot-script.nu, and put it in the nu_scripts/themes/themes folder, you should be able to reproduce this but without the screenshots.

## Preview Theme Script
[preview-theme.nu.txt](https://github.com/nushell/nushell/files/11556242/preview-theme.nu.txt)
## Preview Generate Script
[preview-generate-script.nu.txt](https://github.com/nushell/nushell/files/11556245/preview-generate-script.nu.txt)
## Preview Screenshot Script
[preview-screenshot-script.nu.txt](https://github.com/nushell/nushell/files/11556250/preview-screenshot-script.nu.txt)

**Comment by kubouch** [comment]
So the issue is that
```
def ctxIssue [] { $env.config.color_config.header = blue; ls } 
ctxIssue
```
does not show `ls` output with blue column names, right?

I think the issue is that inside `crates/nu-command/src/viewers/table.rs` the config seems to be fetched by `engine_state.get_config()` which ignores any changes to the environment made in the current parse-eval cycle. These are are only in the Stack until the entire evaluation is over and they're merged into the EngineState, but then it's too late. 

This means that basically any call to `engine_state.get_config()` should be changed to `engine_state.get_config() + combine with $env.config in the Stack`. The only time you'd use plain `engine_state.get_config()` is when you don't have a runtime Stack, i.e., in the parser. We might have such helper already, if not, then it would be probably a good idea to make something like `Stack::get_config(&self, engine_state: &EngineState)`. You can see a similar idea in the `Stack::get_env_var(&self, engine_state: &EngineState)`: We also pass the engine state and look both there and in the stack for the environment variable.

**Comment by fdncred** [comment]
closed by #9304
